### PR TITLE
raft: transfer leader feature

### PR DESCRIPTION
**Pull request by es-chow**
Try to implement transfer leader as per raft thesis `3.10 Leadership transfer extension`.

### CODE REVIEW

**Comment by xiang90** [comment]
@es-chow Can you talk a little bit about your use case or potential use case? I know some one has already done this for etcd/raft. I can ask him to review the code too.

**Comment by es-chow** [comment]
The background for this is that Cockroachdb has an idea to coincide raft leader with range leader, https://github.com/cockroachdb/cockroach/blob/master/docs/design.md#relationship-to-raft-leadership.

This coincide may be achieved by leader transfer.

**Comment by xiang90** [comment]
@bdarnell Can you confirm cockroach need this feature?

File: raft/raft.go [hunk 1]
```diff
@@ -448,6 +452,12 @@ func (r *raft) tickHeartbeat() {
 		r.heartbeatElapsed = 0
 		r.Step(pb.Message{From: r.id, Type: pb.MsgBeat})
 	}
+
+	// If current leader cannot transfer leader in election timeout, become leader again.
```
Comment (bdarnell) [inline raft/raft.go]: Shouldn't this be inside the `r.electionElapsed >= r.electionTimeout` block above at line 440?

File: raft/raft.go [hunk 2]
```diff
@@ -448,6 +452,12 @@ func (r *raft) tickHeartbeat() {
...
+	// As per raft thesis 3.10
+	if r.state == StateTransferLeader {
+		r.becomeLeader()
```
Comment (bdarnell) [inline raft/raft.go]: I think `becomeLeader` does too much: scanning the tail of the log for uncommitted config changes, appending a dummy entry, etc. All we really need here is to reset a few fields including `r.state` and `r.step`, then continue as if our leadership was never interrupted.

**Comment by bdarnell** [comment]
Yes, we could use some sort of orderly leadership handoff to allow us to choose the best leader.

I don't like doing this by introducing a new `StateTransferLeader` that is almost exactly like `StateLeader`, though. I think it would be better to stay in `StateLeader` throughout the process (until the transferee sends us a `MsgVote` and the old leader becomes a follower). In `stepLeader`'s handling of `MsgProp`, we could drop proposals whenever `transferee` is non-zero.

**Comment by xiang90** [comment]
> Yes, we could use some sort of orderly leadership handoff to allow us to choose the best leader.

OK. I will take a closer look at this pr.

File: raft/raft.go (hunk 1 above)
Comment (es-chow) [inline raft/raft.go]: yes, will fix it.

File: raft/raft.go (hunk 2 above)
Comment (es-chow) [inline raft/raft.go]: can fix as per your suggestion.

File: raft/raft.go [hunk 3]
```diff
@@ -180,7 +182,9 @@ type raft struct {
 	tick             func()
 	step             stepFunc
 
-	logger Logger
+	// Leader transfer target.
+	transferee uint64
```
Comment (xiang90) [inline raft/raft.go]: move this to line 162?

**Comment by es-chow** [comment]
@bdarnell, as transferee will be reset to zero after sending out a MsgTimeoutNow to prevent multiple MsgTimeoutNow to send out when pipelining MsgAppResp been received. So if we judge if a MsgProp can be honored by check the value of transferee, it may make a MsgProp being honored just after a MsgTimeoutNow been sent out. Or we have to introduce one more flag to distinguish this.

**Comment by xiang90** [comment]
I believe @dterei has implemented this feature for etcd/raft. So we would love to hear his opinion!

**Comment by xiang90** [comment]
@es-chow I agree with @bdarnell. If the current leader receives a `transfer` it can still be in leader state. But it starts to block all incoming proposals for some timeout and sends all pending entries to the transferee. Adding one or two flags is probably better than introducing a whole new state?

File: raft/node_test.go [hunk 4]
```diff
@@ -42,7 +42,7 @@ func TestNodeStep(t *testing.T) {
 				t.Errorf("%d: cannot receive %s on propc chan", msgt, msgn)
 			}
 		} else {
-			if msgt == raftpb.MsgBeat || msgt == raftpb.MsgHup || msgt == raftpb.MsgUnreachable || msgt == raftpb.MsgSnapStatus || msgt == raftpb.MsgCheckQuorum {
+			if msgt == raftpb.MsgBeat || msgt == raftpb.MsgHup || msgt == raftpb.MsgUnreachable || msgt == raftpb.MsgSnapStatus || msgt == raftpb.MsgCheckQuorum || msgt == raftpb.MsgTransferLeader {
```
Comment (xiang90) [inline raft/node_test.go]: probably it is time for us to make this multiple lines.

**Comment by es-chow** [comment]
@xiang90 , @bdarnell , @dterei, updated as per your comments. PTAL.

File: raft/raft.go [hunk 5]
```diff
@@ -158,7 +158,10 @@ type raft struct {
 
 	// the leader id
 	lead uint64
-
+	// Leader transfer target.
+	transferee uint64
```
Comment (xiang90) [inline raft/raft.go]: // leadTransfree is id of the leader transfer target when its value is not zero.
leadTransfree?

File: raft/raft.go [hunk 6]
```diff
@@ -596,6 +608,11 @@ func stepLeader(r *raft, m pb.Message) {
 			// drop any new proposals.
 			return
 		}
+		if r.transferee != None {
+			r.logger.Debugf("%x ignore MsgProp when TransferLeader in progoress", r.id)
```
Comment (xiang90) [inline raft/raft.go]: ignore -> ignored

when TransferLeader is in progress

File: raft/raft.go [hunk 7]
```diff
@@ -654,6 +671,12 @@ func stepLeader(r *raft, m pb.Message) {
 					// an update before, send it now.
 					r.sendAppend(m.From)
 				}
+				// LeaderTransfer in progress.
+				if m.From == r.transferee && pr.Match == r.raftLog.lastIndex() && !r.timeoutNowSent {
```
Comment (xiang90) [inline raft/raft.go]: abstract this into a function?

File: raft/raft.go [hunk 8]
```diff
@@ -654,6 +671,12 @@ func stepLeader(r *raft, m pb.Message) {
...
+					r.timeoutNowSent = true
+					r.logger.Infof("%x send TimeoutNow in %v to %x after MsgAppResp", r.id, r.state, m.From)
```
Comment (xiang90) [inline raft/raft.go]: %x sent TimeoutNow to %x [state: %d, index: %d]

File: raft/raft.go [hunk 9]
```diff
@@ -689,6 +712,32 @@ func stepLeader(r *raft, m pb.Message) {
 			pr.becomeProbe()
 		}
 		r.logger.Debugf("%x failed to send message to %x because it is unreachable [%s]", r.id, m.From, pr)
+	case pb.MsgTransferLeader:
```
Comment (xiang90) [inline raft/raft.go]: non-leader should filter out this message type and print out debugging info?

File: raft/raft.go [hunk 10]
```diff
@@ -689,6 +712,32 @@ func stepLeader(r *raft, m pb.Message) {
...
+		transferee := m.From
+		if transferee == r.id {
+			// If transfer to myself, noop.
```
Comment (xiang90) [inline raft/raft.go]: remove the comment? it is meaningless with the clear code.

File: raft/raft.go [hunk 11]
```diff
@@ -689,6 +712,32 @@ func stepLeader(r *raft, m pb.Message) {
...
+			if r.transferee == None {
+				r.logger.Debugf("%x transfer leader to myself, noop", r.id)
```
Comment (xiang90) [inline raft/raft.go]: %x is already leader. Ignored transfer request to %x, r.id, r.id

File: raft/raft.go [hunk 12]
```diff
@@ -689,6 +712,32 @@ func stepLeader(r *raft, m pb.Message) {
...
+				return
+			}
+			// Abort previous transfer, transfer to myself agian.
```
Comment (xiang90) [inline raft/raft.go]: debug logging here? remove the comment?

File: raft/raft.go [hunk 13]
```diff
@@ -689,6 +712,32 @@ func stepLeader(r *raft, m pb.Message) {
...
+			r.transferee = None
+			r.timeoutNowSent = false
+			return
+		}
+		// Transfer to third party.
+		r.electionElapsed = 0
```
Comment (xiang90) [inline raft/raft.go]: i feel it worth to doc why we reset electionElapsed.

File: raft/raft.go [hunk 14]
```diff
@@ -689,6 +712,32 @@ func stepLeader(r *raft, m pb.Message) {
...
+		r.transferee = transferee
+		r.logger.Infof("%x try to transfer leader to %x at term %d", r.id, transferee, r.Term)
```
Comment (xiang90) [inline raft/raft.go]: %x starts to transfer leadership to %x at term %d

File: raft/raft.go [hunk 15]
```diff
@@ -689,6 +712,32 @@ func stepLeader(r *raft, m pb.Message) {
...
+		if pr.Match == r.raftLog.lastIndex() {
+			r.timeoutNowSent = true
+			r.send(pb.Message{To: transferee, Type: pb.MsgTimeoutNow})
+			r.logger.Infof("%x send TimeoutNow to %x as it has latest log", r.id, transferee)
```
Comment (xiang90) [inline raft/raft.go]: %x sends TimeoutNow to %x immediately as %x already has up-to-date log.

File: raft/raft.go [hunk 16]
```diff
@@ -689,6 +712,32 @@ func stepLeader(r *raft, m pb.Message) {
...
+		} else {
+			r.timeoutNowSent = false
+			r.sendAppend(transferee)
+			r.logger.Infof("%x send Append to %x as it has match %x less than lastIndex %x", r.id, transferee, pr.Match, r.raftLog.lastIndex())
```
Comment (xiang90) [inline raft/raft.go]: remove line 739? it seems meaningless.

File: raft/raft.go [hunk 17]
```diff
@@ -689,6 +712,32 @@ func stepLeader(r *raft, m pb.Message) {
...
+		// Transfer to third party.
+		r.electionElapsed = 0
+		r.transferee = transferee
+		r.logger.Infof("%x try to transfer leader to %x at term %d", r.id, transferee, r.Term)
+		if pr.Match == r.raftLog.lastIndex() {
+			r.timeoutNowSent = true
+			r.send(pb.Message{To: transferee, Type: pb.MsgTimeoutNow})
+			r.logger.Infof("%x send TimeoutNow to %x as it has latest log", r.id, transferee)
+		} else {
+			r.timeoutNowSent = false
```
Comment (xiang90) [inline raft/raft.go]: remove 737? the default value is false?

File: raft/raft.go [hunk 18]
```diff
@@ -755,6 +804,9 @@ func stepFollower(r *raft, m pb.Message) {
 				r.id, r.raftLog.lastTerm(), r.raftLog.lastIndex(), r.Vote, m.From, m.LogTerm, m.Index, r.Term)
 			r.send(pb.Message{To: m.From, Type: pb.MsgVoteResp, Reject: true})
 		}
+	case pb.MsgTimeoutNow:
+		r.logger.Infof("%x starts election at term %d after received TimeoutNow from %x", r.id, r.Term, m.From)
```
Comment (xiang90) [inline raft/raft.go]: %x received TimeoutNow from %x and starts an election to get leadership.

File: raft/raft.go [hunk 19]
```diff
@@ -848,6 +900,11 @@ func (r *raft) removeNode(id uint64) {
 	if r.maybeCommit() {
 		r.bcastAppend()
 	}
+	// If the removed node is the transferee, then abort the transfering.
+	if r.state == StateLeader && r.transferee == id {
```
Comment (xiang90) [inline raft/raft.go]: make this a func called abortLeaderTransfer?

File: raft/raft.go [hunk 20]
```diff
@@ -444,6 +450,12 @@ func (r *raft) tickHeartbeat() {
 		if r.checkQuorum {
 			r.Step(pb.Message{From: r.id, Type: pb.MsgCheckQuorum})
 		}
+		// If current leader cannot transfer leader in election timeout, become leader again.
+		// As per raft thesis 3.10
+		if r.state == StateLeader && r.transferee != None {
```
Comment (xiang90) [inline raft/raft.go]: abstract this to a func called abortLeaderTransfer?

File: raft/raft_test.go [hunk 21]
```diff
@@ -1911,6 +1911,244 @@ func TestCommitAfterRemoveNode(t *testing.T) {
 	}
 }
 
+func TestNormalLeaderTransfer(t *testing.T) {
```
Comment (xiang90) [inline raft/raft_test.go]: can we be more clear about `Normal`? What does it mean?

Comment (xiang90) [inline raft/raft_test.go]: I would suggest to add a comment to make it clear that we test the code path that the transferee has the most up-to-date log entires when transfer starts.

File: raft/raft_test.go [hunk 22]
```diff
@@ -1911,6 +1911,244 @@ func TestCommitAfterRemoveNode(t *testing.T) {
...
+
+	// Transfer leader to 2.
+	nt.send(pb.Message{From: 2, To: 1, Type: pb.MsgTransferLeader})
+
+	if lead.lead != 2 || lead.transferee != None {
+		t.Fatalf("After first time transferring, leader is %x transferee %v, want leader 2, transferee %v", lead.lead, lead.transferee, None)
+	}
+
+	// After some log replication, transfer leader back to 1.
+	for j := 0; j <= 5; j++ {
```
Comment (xiang90) [inline raft/raft_test.go]: why not use i but j?

File: raft/raft_test.go [hunk 23]
```diff
@@ -1911,6 +1911,244 @@ func TestCommitAfterRemoveNode(t *testing.T) {
...
+	}
+
+	nt.send(pb.Message{From: 1, To: 2, Type: pb.MsgTransferLeader})
+
+	if lead.lead != 1 || lead.transferee != None {
+		t.Fatalf("After second time transferring, leader is %x transferee %v, want leader 1, transferee %v", lead.lead, lead.transferee, None)
+	}
+}
+
+func TestTransferAfterLogReplication(t *testing.T) {
```
Comment (xiang90) [inline raft/raft_test.go]: TestLeaderTransferToSlowFollower

File: raft/raft_test.go [hunk 24]
```diff
@@ -1911,6 +1911,244 @@ func TestCommitAfterRemoveNode(t *testing.T) {
...
+	for i := 0; i < lead.electionTimeout-lead.heartbeatTimeout+1; i++ {
+		lead.tick()
+	}
+
+	if lead.state != StateLeader || lead.lead != 1 || lead.transferee != None {
+		t.Fatalf("Node 1 has state %v lead %v transferee %v, want state %v leader %v transferee %v", lead.state, lead.lead, lead.transferee, StateLeader, 1, None)
+	}
+}
+
+func TestTransferIgnorePropose(t *testing.T) {
```
Comment (xiang90) [inline raft/raft_test.go]: TestLeaderTransferIgnoreProposal (propose to proposal)

File: raft/raft_test.go [hunk 25]
```diff
@@ -1911,6 +1911,244 @@ func TestCommitAfterRemoveNode(t *testing.T) {
...
+	nt := newNetwork(nil, nil, nil)
+	nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgHup})
+
+	nt.isolate(3)
+
+	lead := nt.peers[1].(*raft)
+
+	// Transfer leader to isolated node to let transfer pending, then send propose.
```
Comment (xiang90) [inline raft/raft_test.go]: proposal

File: raft/raft_test.go [hunk 26]
```diff
@@ -1911,6 +1911,244 @@ func TestCommitAfterRemoveNode(t *testing.T) {
...
+	nt.storage[1].Compact(lead.raftLog.applied)
+
+	nt.recover()
+	if lead.prs[3].Match != 1 {
+		t.Fatalf("Node 3 has match %x, want %x", lead.prs[3].Match, 1)
+	}
+
+	// Transfer leader to 3 when node 3 is lack of snapshot.
+	nt.send(pb.Message{From: 3, To: 1, Type: pb.MsgTransferLeader})
+	// Only pb.MsgHeartbeatResp can trigger a snapshot.
```
Comment (xiang90) [inline raft/raft_test.go]: // Send pb.MsgHeartbeatResp to leader to trigger a snapshot for node 3.

File: raft/raft_test.go [hunk 27]
```diff
@@ -1911,6 +1911,244 @@ func TestCommitAfterRemoveNode(t *testing.T) {
...
+	nt.send(pb.Message{From: 3, To: 1, Type: pb.MsgHeartbeatResp})
+
+	if lead.lead != 3 || lead.transferee != None {
+		t.Fatalf("After transferring, leader is %x transferee %v, want leader 3, transferee %v", lead.lead, lead.transferee, None)
+	}
+}
+
+func TestTransferToMyselfOrNonExistNode(t *testing.T) {
```
Comment (xiang90) [inline raft/raft_test.go]: One test func for one test case? 

So TestLeaderTransferToSelf and TestLeaderTransferToNonExistingNode

File: raft/raft_test.go [hunk 28]
```diff
@@ -1911,6 +1911,244 @@ func TestCommitAfterRemoveNode(t *testing.T) {
...
+	nt := newNetwork(nil, nil, nil)
+	nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgHup})
+
+	for j := 0; j <= 5; j++ {
```
Comment (xiang90) [inline raft/raft_test.go]: why do we send 5 proposals in this test case?

File: raft/raft_test.go [hunk 29]
```diff
@@ -1911,6 +1911,244 @@ func TestCommitAfterRemoveNode(t *testing.T) {
...
+	// Transfer leader to isolated node, wait for timeout.
+	nt.send(pb.Message{From: 3, To: 1, Type: pb.MsgTransferLeader})
+	if lead.transferee == None {
+		t.Errorf("Waiting transferring, transferee = %v, want %v", lead.transferee, 3)
+	}
+	for i := 0; i < lead.heartbeatTimeout+1; i++ {
+		lead.tick()
+	}
+	if lead.transferee == None {
+		t.Errorf("Waiting transferring, transferee = %v, want %v", lead.transferee, 3)
```
Comment (xiang90) [inline raft/raft_test.go]: t.Fatalf (i do not think we can continue this test if we failed here)

error message: want transferring: transferee = %v, want %v.

File: raft/raft_test.go [hunk 30]
```diff
@@ -1911,6 +1911,244 @@ func TestCommitAfterRemoveNode(t *testing.T) {
...
+
+	// Transfer leader to isolated node, wait for timeout.
+	nt.send(pb.Message{From: 3, To: 1, Type: pb.MsgTransferLeader})
+	if lead.transferee == None {
+		t.Errorf("Waiting transferring, transferee = %v, want %v", lead.transferee, 3)
+	}
+	for i := 0; i < lead.heartbeatTimeout+1; i++ {
+		lead.tick()
+	}
+	if lead.transferee == None {
```
Comment (xiang90) [inline raft/raft_test.go]: if lead.transferee != 3

File: raft/raft_test.go [hunk 31]
```diff
@@ -1911,6 +1911,244 @@ func TestCommitAfterRemoveNode(t *testing.T) {
...
+		t.Errorf("Waiting transferring, transferee = %v, want %v", lead.transferee, 3)
+	}
+
+	for i := 0; i < lead.electionTimeout-lead.heartbeatTimeout+1; i++ {
+		lead.tick()
+	}
+
+	if lead.state != StateLeader || lead.lead != 1 || lead.transferee != None {
```
Comment (xiang90) [inline raft/raft_test.go]: shall we also compare lead.TimeoutNow?

File: raft/raft_test.go [hunk 32]
```diff
@@ -1911,6 +1911,244 @@ func TestCommitAfterRemoveNode(t *testing.T) {
...
+	for j := 0; j <= 5; j++ {
+		nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgProp, Entries: []pb.Entry{{}}})
+	}
+
+	if lead.prs[1].Match != 1 {
+		t.Fatalf("Node 1 has match %x, want %x", lead.prs[1].Match, 1)
+	}
+}
+
+func TestTransferReceiveHighTermVote(t *testing.T) {
```
Comment (xiang90) [inline raft/raft_test.go]: TestLeaderTransferReceiveHigherTermVote

File: raft/rawnode_test.go [hunk 33]
```diff
@@ -33,7 +33,8 @@ func TestRawNodeStep(t *testing.T) {
 		msgt := raftpb.MessageType(i)
 		err = rawNode.Step(raftpb.Message{Type: msgt})
 		// LocalMsg should be ignored.
-		if msgt == raftpb.MsgBeat || msgt == raftpb.MsgHup || msgt == raftpb.MsgUnreachable || msgt == raftpb.MsgSnapStatus {
+		if msgt == raftpb.MsgBeat || msgt == raftpb.MsgHup || msgt == raftpb.MsgUnreachable ||
```
Comment (xiang90) [inline raft/rawnode_test.go]: use isLocalMessage?

File: raft/rawnode.go [hunk 34]
```diff
@@ -226,3 +226,8 @@ func (rn *RawNode) ReportSnapshot(id uint64, status SnapshotStatus) {
 
 	_ = rn.raft.Step(pb.Message{Type: pb.MsgSnapStatus, From: id, Reject: rej})
 }
+
+// TransferLeader makes raft try to transfer leader to transferee.
+func (rn *RawNode) TransferLeader(transferee uint64) {
```
Comment (xiang90) [inline raft/rawnode.go]: TransferLeader tries to transfer leadership to the given transferee.

File: raft/raft_test.go [hunk 35]
```diff
@@ -1911,6 +1911,244 @@ func TestCommitAfterRemoveNode(t *testing.T) {
...
+	}
+
+	nt.send(pb.Message{From: 2, To: 2, Type: pb.MsgHup, Index: 1, Term: 2})
+
+	if lead.state != StateFollower || lead.lead != 2 || lead.transferee != None {
+		t.Fatalf("Node 1 has state %v lead %v transferee %v, want state %v leader %v transferee %v", lead.state, lead.lead, lead.transferee, StateFollower, 2, None)
+	}
+}
+
+func TestTransferRemoveNode(t *testing.T) {
```
Comment (xiang90) [inline raft/raft_test.go]: TestLeaderTransferRemoveNode

File: raft/raft_test.go [hunk 36]
```diff
@@ -1911,6 +1911,244 @@ func TestCommitAfterRemoveNode(t *testing.T) {
...
+	}
+
+	lead.removeNode(3)
+
+	if lead.state != StateLeader || lead.lead != 1 || lead.transferee != None {
+		t.Fatalf("Node 1 has state %v lead %v transferee %v, want state %v leader %v transferee %v", lead.state, lead.lead, lead.transferee, StateLeader, 1, None)
+	}
+}
+
+func TestTransferBack(t *testing.T) {
```
Comment (xiang90) [inline raft/raft_test.go]: TestLeaderTransferBack.

we need to add comment to this test about what it does. it has two steps and it is not clear from the function name.

**Comment by xiang90** [comment]
@es-chow Looks good overall. Please fix the comments. Add LeaderTransfere prefix to all tests. Thanks!

**Comment by es-chow** [comment]
@xiang90 , @bdarnell, updated as per the comments. PTAL.

File: raft/raft.go [hunk 37]
```diff
@@ -720,6 +773,10 @@ func stepCandidate(r *raft, m pb.Message) {
 		case len(r.votes) - gr:
 			r.becomeFollower(r.Term, None)
 		}
+	case pb.MsgTransferLeader:
```
Comment (xiang90) [inline raft/raft.go]: can you move these to line 555. (same as how we handle m.Type == pb.MsgHup )

File: raft/raft.go [hunk 38]
```diff
@@ -755,6 +812,11 @@ func stepFollower(r *raft, m pb.Message) {
 				r.id, r.raftLog.lastTerm(), r.raftLog.lastIndex(), r.Vote, m.From, m.LogTerm, m.Index, r.Term)
 			r.send(pb.Message{To: m.From, Type: pb.MsgVoteResp, Reject: true})
 		}
+	case pb.MsgTimeoutNow:
```
Comment (xiang90) [inline raft/raft.go]: can you move these to line 555. (same as how we handle m.Type == pb.MsgHup )

File: raft/raft.go [hunk 39]
```diff
@@ -755,6 +812,11 @@ func stepFollower(r *raft, m pb.Message) {
...
+		r.logger.Infof("%x [term %d] received MsgTimeoutNow from %x and starts an election to get leadership.", r.id, r.Term, m.From)
+		r.campaign()
+	case pb.MsgTransferLeader:
```
Comment (xiang90) [inline raft/raft.go]: can you move these to line 555. (same as how we handle m.Type == pb.MsgHup )

File: raft/raft.go [hunk 40]
```diff
@@ -596,6 +608,11 @@ func stepLeader(r *raft, m pb.Message) {
 			// drop any new proposals.
 			return
 		}
+		if r.leadTransferee != None {
```
Comment (xiang90) [inline raft/raft.go]: I believe @bdarnell wont want to see this to happen in cockroach... at least, we should add a todo to buffer these proposals?

**Comment by xiang90** [comment]
LGTM. Defer to @bdarnell or anyone from cockroach team.

**Comment by ngaut** [comment]
@xiang90 I'd like to see this happen too. And i will port the feature to TiKV later.

File: raft/raft.go (hunk 40 above)
Comment (bdarnell) [inline raft/raft.go]: It's fine with me to drop proposals in this case (just like we already drop proposals on candidates or followers when `r.lead` is `None`).

File: raft/raft.go [hunk 41]
```diff
@@ -654,6 +671,11 @@ func stepLeader(r *raft, m pb.Message) {
 					// an update before, send it now.
 					r.sendAppend(m.From)
 				}
+				// Transfer leadership is in progress.
+				if m.From == r.leadTransferee && pr.Match == r.raftLog.lastIndex() && !r.timeoutNowSent {
+					r.logger.Infof("%x sent MsgTimeoutNow to %x after received MsgAppResp", r.id, m.From)
```
Comment (bdarnell) [inline raft/raft.go]: If we assume that the transport (usually) preserves order, we could send the `MsgTimeoutNow` as soon as we have sent a `MsgApp` containing the last index, without waiting for the `MsgAppResp`. This is similar to the way we optimistically increase `Progress.Next` when sending `MsgApp`.

**Comment by bdarnell** [comment]
LGTM. Thanks @es-chow!

**Comment by xiang90** [comment]
@es-chow Can you fix the last few comments? We are getting really close to get this merged! Thanks!

File: raft/raft.go (hunk 41 above)
Comment (es-chow) [inline raft/raft.go]: @bdarnell, @xiang90, I feel we cannot send `MsgTimeoutNow` as soon as we have sent a ` MsgApp`  containing the last index in all cases, if we add that code and remove the code to send `MsgTimeoutNow ` when a `MsgAppResp` is received, the test case `TestLeaderTransferToSlowFollower` cannot be passed as node 3 will reject that `MsgApp` just before the `MsgTimeoutNow` triggered by `pb.MsgTransferLeader` as its log is not up-to-date, but the following `MsgTimeoutNow` will trigger node3 to start an election which also will fail.

1. Maybe one way is we keep the code which send `MsgTimeoutNow` as receiving `MsgAppResp`, but add one optimistic sending `MsgTimeoutNow` only when we optimistically update the `pr.Next`, but this may also have the risk that peer may reject that `MsgApp` and  `MsgTimeoutNow` will make the group unavailable for one election timeout. And also we have to keep the `timeoutNowSent` flag.

2. We keep the current code as we can also remove the `timeoutNowSent` flag. This is the updated commit.

Comment (xiang90) [inline raft/raft.go]: I feel this is an optimization. We can think this more and do it in another pr. @es-chow Can you set up another issue to track this potential optimization?

Comment (bdarnell) [inline raft/raft.go]: Yeah, this optimization can wait. One thing we could do to avoid the scenario in which the follower rejects the MsgApp and then starts an election that fails is to include the leader's last log index in the MsgTimeoutNow. The follower would only start campaigning if it matches the index; if it doesn't then it knows it can't win.

File: raft/raft.go [hunk 42]
```diff
@@ -687,6 +710,37 @@ func stepLeader(r *raft, m pb.Message) {
...
+				return
+			}
+			r.abortLeaderTransfer()
+			r.logger.Infof("%x [term %d] abort transfer leadership to %x", r.id, r.Term, lastLeadTransferee)
+		}
+		if leadTransferee == r.id {
+			if lastLeadTransferee == None {
+				r.logger.Debugf("%x is already leader. Ignored transfer leadership to %x", r.id, r.id)
+			} else {
+				r.logger.Debugf("%x abort transfer leadership to %x, transfer to current leader %x.", r.id, lastLeadTransferee, r.id)
```
Comment (tiancaiamao) [inline raft/raft.go]: it seems we never run to this line, otherwise r.abortLeaderTransfer() should be add.

Comment (xiang90) [inline raft/raft.go]: I agree /cc @es-chow can you also take a look?
